pkg/local_object_storage/metabase/metadata.go

```diff
@@ -0,0 +1,796 @@
+		return nil, errors.New("wrong attribute")
+	}
+	var res SearchCursor
+	res.ValIDOff = 1 + len(primAttr) + len(utf8Delimiter)
+	if len(b[res.ValIDOff:]) <= oid.Size {
+		return nil, errors.New("missing value")
+	}
+	res.Key = b
+	return &res, nil
+}
+
+// Search selects up to count container's objects from the given container
+// matching the specified filters.
+func (db *DB) Search(cnr cid.ID, fs object.SearchFilters, attrs []string, cursor *SearchCursor, count uint16) ([]client.SearchResultItem, *SearchCursor, error) {
+	if blindlyProcess(fs) {
+		return nil, nil, nil
+	}
+	var res []client.SearchResultItem
+	var newCursor *SearchCursor
+	var err error
+	if len(fs) == 0 {
+		res, newCursor, err = db.searchUnfiltered(cnr, cursor, count)
+	} else {
+		res, newCursor, err = db.search(cnr, fs, attrs, cursor, count)
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+	return res, newCursor, nil
+}
+
+func (db *DB) search(cnr cid.ID, fs object.SearchFilters, attrs []string, cursor *SearchCursor, count uint16) ([]client.SearchResultItem, *SearchCursor, error) {
+	var res []client.SearchResultItem
+	var newCursor *SearchCursor
+	err := db.boltDB.View(func(tx *bbolt.Tx) error {
+		mb := tx.Bucket(metaBucketKey(cnr))
+		if mb == nil {
+			return nil
+		}
+		var err error
+		res, newCursor, err = db.searchInBucket(mb, fs, attrs, cursor, count)
+		return err
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("view BoltDB: %w", err)
+	}
+	return res, newCursor, nil
+}
+
+func (db *DB) searchInBucket(metaBkt *bbolt.Bucket, fs object.SearchFilters, attrs []string,
+	cursor *SearchCursor, count uint16) ([]client.SearchResultItem, *SearchCursor, error) {
+	// TODO: make as much as possible outside the Bolt tx
+	primMatcher := fs[0].Operation()
+	intPrimMatcher := isNumericOp(primMatcher)
+	notPresentPrimMatcher := primMatcher == object.MatchNotPresent
+	primAttr := fs[0].Header() // attribute emptiness already prevented
+	var primSeekKey, primSeekPrefix []byte
+	var prevResOID, prevResPrimVal []byte
+	if notPresentPrimMatcher {
+		if cursor != nil {
+			primSeekKey = cursor.Key
+		} else {
+			primSeekKey = make([]byte, 1)
+		}
+		primSeekKey[0] = metaPrefixID
+		primSeekPrefix = primSeekKey[:1]
+	} else if cursor != nil {
+		primSeekKey = cursor.Key
+		if intPrimMatcher {
+			primSeekKey[0] = metaPrefixAttrIDInt
+		} else {
+			primSeekKey[0] = metaPrefixAttrIDPlain
+		}
+		primSeekPrefix = primSeekKey[:cursor.ValIDOff]
+		valID := cursor.Key[cursor.ValIDOff:]
+		prevResPrimVal, prevResOID = valID[:len(valID)-oid.Size], valID[len(valID)-oid.Size:]
+	} else {
+		if primMatcher == object.MatchStringEqual || primMatcher == object.MatchCommonPrefix ||
+			primMatcher == object.MatchNumGT || primMatcher == object.MatchNumGE {
+			var err error
+			if primSeekKey, primSeekPrefix, err = seekKeyForAttribute(primAttr, fs[0].Value()); err != nil {
+				return nil, nil, fmt.Errorf("invalid primary filter value: %w", err)
```

Comment:
Shouldn't this be checked _before_ going into `searchInBucket`?

Reply:
can be, dont think this is sufficient for now